libsrc/iofuncs/man3/im_init.3
.TH IM_INIT 3 "11 April 1990"
.SH NAME
im_init \- make an IMAGE descriptor
.SH SYNOPSIS
#include <vips/vips.h>

IMAGE *im_init( char *filename )
.SH DESCRIPTION
im_init(3) allocates space for an IMAGE descriptor, sets all fields to
initial values, and returns the descriptor. A copy is made of the filename
argument. The IMAGE returned by im_init(3) should be passed to im_close(3) to
free it.

This function is used internally by VIPS and should not be called by users.
.SH SEE ALSO
im_mmapin(3), im_openin(3), im_setbuf(3), im_close(3).
.SH COPYRIGHT
Birkbeck College and the National Gallery (c) 1994
.SH AUTHOR
N. Dessipris \- 11/04/1990
.br
J.Cupitt \- 23/2/94


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/iofuncs/im_init.c
/* @(#) Initialise the IMAGE to impossible startup values. Install the
 * @(#) filename.
 * @(#) 
 * @(#) IMAGE *im_init( char *filename )
 *
 * Copyright: Nicos Dessipris & Kirk Martinez, 1990
 * Written on: 13/02/1990
 * Modified on : 3/6/92 22/2/93
 * 15/4/93 J.Cupitt
 *	- init for partial image buffers added
 *	- init for type field
 *	- filename added
 * 10/5/93 J.Cupitt
 *	- allocates space for IMAGE too, and returns new data
 * 23/2/94 JC
 *	- ANSIfied, man page revised
 * 16/8/94 JC
 *	- evalend callbacks added
 * 28/11/94 JC
 *	- new compression fields added, thr added
 * 24/10/95 JC
 *	- now tracks open images ... see also im_close() and debug.c
 * 1/12/04 JC
 *	- added an im_init_world() to help old progs
 * 30/9/05
 * 	- added sizeof_header
 * 2/1/07
 * 	- init magic
 * 	- init history_list
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

/*
#define DEBUG_NEW
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <vips/vips.h>
#include <vips/debug.h>
#include <vips/thread.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Make a new IMAGE structure, set fields to an initial state. We set the
 * filename field only.
 */
IMAGE *
im_init( const char *filename )
{	
	IMAGE *im;

	/* Pass in a nonsense name for argv0 ... this init world is only here
	 * for old programs which are missing an im_init_world() call. We must
	 * have threads set up before we can process.
	 */
	if( im_init_world( "vips" ) )
		im_error_clear();

	if( !(im = IM_NEW( NULL, IMAGE )) )
		return( NULL );

#ifdef DEBUG_NEW
	printf( "im_init: new IMAGE 0x%p, \"%s\"\n", im, filename );
#endif /*DEBUG_NEW*/

	im->Xsize = -1;
	im->Ysize = -1;
	im->Bands = -1;
	im->Bbits = -1;
	im->BandFmt = -1;
	im->Coding = -1;
	im->Type = -1;
	im->Xres = 1.0;
	im->Yres = 1.0;
	im->Length = 0;
	im->Compression = 0;
	im->Level = 0;
	im->Xoffset = 0;
	im->Yoffset = 0;

	im->Hist = NULL;

	im->data = NULL;
	im->time = NULL;
	im->kill = 0;

	im->dtype = IM_NONE;
	im->fd = -1;
	im->baseaddr = NULL;
	im->length = 0;
	im->closefns = NULL;
	im->evalfns = NULL;
	im->evalendfns = NULL;
	im->closing = 0;
	im->close_pending = 0;

	/* Default to native order.
	 */
	im->magic = im_amiMSBfirst() ?  IM_MAGIC_SPARC : IM_MAGIC_INTEL;

	im->start = NULL;
	im->generate = NULL;
	im->stop = NULL;
	im->client1 = NULL;
	im->client2 = NULL;
	im->sslock = g_mutex_new();
	im->regions = NULL;
	im->dhint = IM_SMALLTILE;

	im->Meta = NULL;
	im->Meta_traverse = NULL;

	/* Default to the VIPS header size. Can be changed later.
	 */
	im->sizeof_header = IM_SIZEOF_HEADER;

	im->windows = NULL;

	im->parents = NULL;
	im->children = NULL;
	im->serial = 0;

	im->history_list = NULL;

	if( !(im->filename = im_strdup( NULL, filename )) ) {
		im_close( im );
		return( NULL );
	}

	im__open_images = g_slist_prepend( im__open_images, im );

	return( im );
}
